File: lib/domain/entities/zones.dart
class SafetyZoneSchedule {
  const SafetyZoneSchedule({
    required this.id,
    required this.zoneId,
    required this.weekday,
    required this.startTime,
    required this.endTime,
  });

  final String id;
  final String zoneId;

  /// 0 = Monday … 6 = Sunday.
  final int weekday;
  final String startTime;
  final String endTime;

  bool get crossesMidnight => endTime.compareTo(startTime) <= 0;

  factory SafetyZoneSchedule.fromJson(Map<String, dynamic> json) {
    return SafetyZoneSchedule(
      id: json['id'] as String,
      zoneId: json['zone_id'] as String,
      weekday: json['weekday'] as int,
      startTime: json['start_time'] as String,
      endTime: json['end_time'] as String,
    );
  }
}

class SafetyZone {
  const SafetyZone({
    required this.id,
    required this.youngPersonId,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.radiusMeters,
    required this.isActive,
    required this.accuracyToleranceMeters,
    required this.minExitDurationSeconds,
    this.schedules = const [],
    this.scheduleActiveNow = false,
    this.insideOnLastFix = false,
  });

  final String id;
  final String youngPersonId;
  final String name;
  final double latitude;
  final double longitude;
  final double radiusMeters;
  final bool isActive;
  final double accuracyToleranceMeters;
  final int minExitDurationSeconds;
  final List<SafetyZoneSchedule> schedules;
  final bool scheduleActiveNow;
  final bool insideOnLastFix;

  factory SafetyZone.fromJson(Map<String, dynamic> json) {
    return SafetyZone(
      id: json['id'] as String,
      youngPersonId: json['young_person_id'] as String,
      name: json['name'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      radiusMeters: (json['radius_meters'] as num).toDouble(),
      isActive: json['is_active'] as bool,
      accuracyToleranceMeters: (json['accuracy_tolerance_meters'] as num).toDouble(),
      minExitDurationSeconds: json['min_exit_duration_seconds'] as int,
      scheduleActiveNow: json['schedule_active_now'] as bool? ?? false,
      insideOnLastFix: json['inside_on_last_fix'] as bool? ?? false,
      schedules: (json['schedules'] as List<dynamic>? ?? [])
          .map((item) => SafetyZoneSchedule.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

class RiskZone {
  const RiskZone({
    required this.id,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.radiusMeters,
    required this.isActive,
    required this.incidentCount,
    this.typicalStartHour,
    this.typicalEndHour,
    this.isHourActiveNow = false,
  });

  final String id;
  final String name;
  final double latitude;
  final double longitude;
  final double radiusMeters;
  final bool isActive;
  final int incidentCount;
  final int? typicalStartHour;
  final int? typicalEndHour;
  final bool isHourActiveNow;

  factory RiskZone.fromJson(Map<String, dynamic> json) {
    return RiskZone(
      id: json['id'] as String,
      name: json['name'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      radiusMeters: (json['radius_meters'] as num).toDouble(),
      isActive: json['is_active'] as bool,
      incidentCount: json['incident_count'] as int,
      typicalStartHour: json['typical_start_hour'] as int?,
      typicalEndHour: json['typical_end_hour'] as int?,
      isHourActiveNow: json['is_hour_active_now'] as bool? ?? false,
    );
  }
}

class RiskIncident {
  const RiskIncident({
    required this.id,
    required this.title,
    required this.latitude,
    required this.longitude,
    required this.occurredAt,
    required this.count,
    this.riskZoneId,
    this.description,
    this.source,
  });

  final String id;
  final String? riskZoneId;
  final String title;
  final String? description;
  final double latitude;
  final double longitude;
  final DateTime occurredAt;
  final String? source;
  final int count;

  factory RiskIncident.fromJson(Map<String, dynamic> json) {
    return RiskIncident(
      id: json['id'] as String,
      riskZoneId: json['risk_zone_id'] as String?,
      title: json['title'] as String,
      description: json['description'] as String?,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      occurredAt: DateTime.parse(json['occurred_at'] as String),
      source: json['source'] as String?,
      count: json['count'] as int,
    );
  }
}

File: backend/app/models/zones.py
from __future__ import annotations

from datetime import datetime, time
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Time, UniqueConstraint, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.mixins import TimestampMixin


class SafetyZone(TimestampMixin, Base):
    """Zone de sécurité (école, maison…) liée à un jeune, avec rayon et plages horaires."""

    __tablename__ = "safety_zones"

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    young_person_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("young_persons.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(80), nullable=False)
    latitude: Mapped[float] = mapped_column(Float, nullable=False)
    longitude: Mapped[float] = mapped_column(Float, nullable=False)
    radius_meters: Mapped[float] = mapped_column(Float, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    # GPS noise must not create a critical alert on its own.
    accuracy_tolerance_meters: Mapped[float] = mapped_column(Float, nullable=False, default=40.0)
    min_exit_duration_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=90)

    young_person: Mapped[YoungPerson] = relationship("YoungPerson", back_populates="safety_zones")
    schedules: Mapped[list[SafetyZoneSchedule]] = relationship(
        "SafetyZoneSchedule",
        back_populates="zone",
        cascade="all, delete-orphan",
    )


class SafetyZoneSchedule(Base):
    """Plage jour + heures. Si end_time <= start_time, la fenêtre traverse minuit (ex. 18:00 → 07:00)."""

    __tablename__ = "safety_zone_schedules"

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    zone_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("safety_zones.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    weekday: Mapped[int] = mapped_column(Integer, nullable=False)  # 0 = lundi … 6 = dimanche
    start_time: Mapped[time] = mapped_column(Time, nullable=False)
    end_time: Mapped[time] = mapped_column(Time, nullable=False)

    zone: Mapped[SafetyZone] = relationship("SafetyZone", back_populates="schedules")


class RiskZone(TimestampMixin, Base):
    """Zone à risque (prévention), distincte d'une zone de sécurité."""

    __tablename__ = "risk_zones"

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    name: Mapped[str] = mapped_column(String(80), nullable=False)
    latitude: Mapped[float] = mapped_column(Float, nullable=False)
    longitude: Mapped[float] = mapped_column(Float, nullable=False)
    radius_meters: Mapped[float] = mapped_column(Float, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    incident_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    typical_start_hour: Mapped[int | None] = mapped_column(Integer, nullable=True)
    typical_end_hour: Mapped[int | None] = mapped_column(Integer, nullable=True)


class RiskZoneOccupancy(TimestampMixin, Base):
    """Présence dans une zone à risque, pour n'alerter qu'à l'entrée."""

    __tablename__ = "risk_zone_occupancies"
    __table_args__ = (UniqueConstraint("young_person_id", "risk_zone_id", name="uq_risk_occupancy_young_zone"),)

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    young_person_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("young_persons.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    risk_zone_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("risk_zones.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    is_inside: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)


class GeofenceOccupancy(TimestampMixin, Base):
    """État d'occupation d'une zone, pour exiger une présence puis une durée hors zone."""

    __tablename__ = "geofence_occupancies"
    __table_args__ = (UniqueConstraint("young_person_id", "zone_id", name="uq_occupancy_young_zone"),)

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    young_person_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("young_persons.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    zone_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("safety_zones.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    observed_inside: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_inside: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    candidate_exit_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

